infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/receive/LogReplicationSinkManager.java

```diff
@@ -226,49 +256,89 @@ public LogReplicationEntry receive(LogReplicationEntry message) {
             return null;
         }
 
+        return processReceivedMessage(message);
+    }
+
+    /**
+     * Process received (valid) message depending on the current rx state (LOG_ENTRY_SYNC or SNAPSHOT_SYNC)
+     *
+     * @param message received message
+     * @return ack
+     */
+    private LogReplicationEntry processReceivedMessage(LogReplicationEntry message) {
         if (rxState.equals(RxState.LOG_ENTRY_SYNC)) {
             return logEntrySinkBufferManager.processMsgAndBuffer(message);
         } else {
-            return snapshotSinkBufferManager.processMsgAndBuffer(message);
+            LogReplicationEntry ack = snapshotSinkBufferManager.processMsgAndBuffer(message);
+            if (ack.getMetadata().getMessageMetadataType() == MessageType.SNAPSHOT_REPLICATED) {
+                long lastStartedSnapshotSync = logReplicationMetadataManager.getLastSnapStartTimestamp();
+                // Verify this snapshot ACK corresponds to the last initialized/valid snapshot sync
+                // as a previous one could have been canceled but still processed due to messages being out of order
+                if (ack.getMetadata().getSnapshotTimestamp() == lastStartedSnapshotSync) {
+                    // Notify end of snapshot sync. This is a blocking call.
```

Comment:
While doing snapshot transfer, all message has the same snapshottimestamp. May check the metadata's seqNum + 1 = ack's seqNum.

Reply:
Make it more clear, we start a snapshot transfer for snapshot timestamp 10, 
the first message is snapshot_start with seqNum= -1 and snapshot timestamp 10,
the second message is snapshot_msg with seqNum=0, and snapshot timestamp 10.
the last one is snapshot_end with seqNum=1, and snapshot timestamp 10.
So all message has the same snapshot timestamp, but different SeqNum. 
May check the metadata:
if snapshot appliedDone == snapshottimestamp then it is done. make it more robust the ack's seqNum is seqNum + 1 (seqNum from metadata)

Reply:
If there are leadership change during full snapshot transfer, we need to consider the following steps:
1. older leader make a call to set the trim holder.
2. leadership. change at the standby cluster, the new leader will re-run the full cycle of snapshot sync by reset the persisted metadata. // this will block the old leader to clear the trim holder.
3. the new leader set the trim holder again. 

If we guarantee that reset the persisted metadata first, then put trim holder, may it be correct?